src/coreclr/jit/abi.h

```diff
@@ -0,0 +1,53 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#pragma once
+
+class ABIPassingSegment
+{
+    regNumber m_register    = REG_NA;
+    unsigned  m_stackOffset = 0;
+
+public:
+    bool IsPassedInRegister() const;
+    bool IsPassedOnStack() const;
+
+    // Start offset of the segment within the parameter/argument. For example, a struct like { int32_t x; uint64_t y }
+    // may have two segments
+    // 1. Register(Offset=0, Type=TYP_INT, Size=4, Register=REG_ESI)
+    // 2. Register(Offset=8, Type=TYP_LONG, Size=8, Register=REG_EDI)
+    // on some ABIs, where the size of the first segment is not sufficient to
+    // compute the offset of the second.
+    unsigned Offset = 0;
+    // Size of the segment being passed.
+    unsigned Size = 0;
+
+    // If this segment is passed in a register, return the particular register.
+    regNumber GetRegister() const;
+
+    // If this segment is passed on the stack then return the particular stack
+    // offset, relative to the first stack argument's offset.
+    unsigned GetStackOffset() const;
+
+    static ABIPassingSegment InRegister(regNumber reg, unsigned offset, unsigned size);
+    static ABIPassingSegment OnStack(unsigned stackOffset, unsigned offset, unsigned size);
+};
+
+struct ABIPassingInformation
+{
+    // The number of segments used to pass the value. Examples:
+    // - On x86, TYP_LONG can be passed in two registers, resulting in two
+    // register segments
+    // - On SysV x64, structs can be passed in two registers, resulting in two
+    // register segments
+    // - On arm64/arm32, HFAs can be passed in up to four registers, giving
+    // four register segments
```

Comment:
This should also include HVAs, which looks like we might be missing handling around:

[godbolt HFA/HVA for Arm64](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIAOxcpK4AMngMmAByPgBGmMQSAMykAA6oCoRODB7evgFBaRmOAmER0SxxCVzJdpgOWUIETMQEOT5%2BgbaY9sUMjc0EpVGx8Um2TS1teZ0KE4PhwxWj1QCUtqhexMjsHOaJ4cjeWADUJoluzSwA%2BhECAHQIZ9gmGgCCL6%2BzxF4OxwASADFXqd/BZjlRaKgmARjqpSMcAJ7wgBe8IA7mcwSZ/AARTEfL4/GF/ABqwOxYIhUIIiTMqkkVxhcMRKPRmJBeMSVjeHyp0OOAFkIIDgQgqEwVh8KR9jrLjsRMARNgxjmKmHdVKdLKrxXcEVqwWq7siDTr1RiuVLcVK3nyaXSGTDIsKyaqAG4Sq3c4FyhVK4gqj3odAARyuVFpECDofDkejYYjZggCA9GvhKfVCLW7vVyOzGbuaMllp5uI4a1onAArLw/BwtKRUJw3NZrMcFBstpgtYkeKQCJpy2sANYgKtcO5mLgANmn/ir064ZgAHFwNMuAJz6TiSWuDxucXgKEAafuDtZwWAwRAoVAsFJ0eLkShoe%2BPhKXN3LgC0h0MwG/AhvgYYc%2BDoAh4mPCAYn3GJwmaBFOD7ODmGIBEAHkYm0OoB24XhXzYQR0IYWhEPrXgsBiLxgAuWhaGPPDSCwFh/3EcimLwBV6jdTAGIbTBVDqLwIKQ3hwggyt2NoPAYmIBCPCwfcgLwFhRLWCEmGABQSTwTA0XQlJGFEmRBBEMR2CkEz5CUNR910IIDCMFBW0sfQZOPSA1lQFJegY790MSXhUB44hiDwLAPIgNZanqZwIFcKY/CCUIFnKSo9EKTIBESjL0iyhghjS5Yuh6Bo5hyoIYt6foWkKkYqnGAYKsa2rUvqiRos7bYOu3Dga1IOsGybDhjk/H8mGOP8jGOCAgK8ECVhm3BCBIHsuBWXhcK0FYRxASRlzufxl0SWlJESZdAk3Jdp163dSFUqtT0GoLD1sE8z3Ii9r2TW833oMgKAgV8H3%2BkA3WQFIUiuN0uA3K4DAg2YrkuadpBoWgIOIKCYPYlCEOMvG0Mw7CHGMgjGAIYjSP3SjqNo%2BjjOY1idgbfAuMcHi%2BN4AShJExjxO6fdpNk%2BSMBZzawtUvD1IMLSdL0gyjMY/hTNEcRLJV6yVHUdjdDMfR/2cyxrDcmJIq8nysj89CzCCkKwoi%2BBou6HCshcBh3E8do9BSsp2oKPLemazLejqpYGqqsqmq9vJKpd2K%2BjmMP0sq8qY6Slr5j98Oeo7TZuvW3r%2Bueg8RpRyRjhYBRweOGGNzuBHeJhCBlqIYg1o2j7trWBBMCYLAEii27eAep792Go93q2odSFHfbDuO07zsujdrt6wKBvH17p523rbc39iJ67meQoyZxJCAA)

[godbolt HVA/HVA for x64](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:67,endLineNumber:13,positionColumn:67,positionLineNumber:13,selectionStartColumn:67,selectionStartLineNumber:13,startColumn:67,startLineNumber:13),source:'%23include+%3Cxmmintrin.h%3E%0A%0Astruct+HFA+%7B+float+x,+y,+z,+w%3B+%7D%3B%0Astruct+HVA+%7B+__m128+x,+y,+z,+w%3B+%7D%3B%0A%0Afloat+M(HFA+hfa)%0A%7B%0A++++return+hfa.x+%2B+hfa.y+%2B+hfa.z+%2B+hfa.w%3B%0A%7D%0A%0A__m128+N(HVA+hva)%0A%7B%0A++++return+_mm_add_ps(_mm_add_ps(_mm_add_ps(hva.x,+hva.y),+hva.z),+hva.w)%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:51.21667561281089,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((g:!((h:compiler,i:(compiler:clang_trunk,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'-O3',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+clang+(trunk)+(Editor+%231)',t:'0')),k:48.783324387189126,l:'4',m:50,n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:vcpp_v19_latest_x64,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:2,lang:c%2B%2B,libs:!(),options:'-O2',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x64+msvc+v19.latest+(Editor+%231)',t:'0')),header:(),l:'4',m:50,n:'0',o:'',s:0,t:'0')),k:48.783324387189126,l:'3',n:'0',o:'',t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

Reply:
These are just examples to indicate examples of the representation. The intention of this PR is not to make any functional changes.